### DISCUSSION TITLE
New calculated (subtraction) column in existing table

### DISCUSSION THREAD

**Discussion by compasssites** [supabase/supabase#5717] [Questions]
Can someone please help me to create a new column to a existing table, where in the new colum (calc) should show the subtraction value of 2 existing columns (one and two). Something like (total stock-currently used stock) = Present stock in hand.

I tried to create a generated column (stored) but it works well for **new rows**. When i try to alter the values of existing rows, it naturally gives the error "column "calc" can only be updated to DEFAULT".

How do i get a column to show simple calculations like in Airtable/any spreadsheet. Can there be additional formula field type? Or am i missing it?

Thanks.

**Comment by GaryAustin1** [comment]
Edit below is not true for POSTGRES just mysql...
You can add a generated column without stored and it just gets calculated from other columns when selected.
ALTER TABLE tablename ADD COLUMN calc int8  GENERATED ALWAYS AS (col1 - col2)

**Comment by compasssites** [comment]
Thanks @GaryAustin1, but it keeps throwing `syntax error at or near ";"`. Cant get the desired result.

I even tried without SORTED at dbfiddle, it doesnt go through. 

I tried 
CREATE TABLE testcalc (
    one numeric,
    two numeric,
    calc numeric GENERATED ALWAYS AS (one-two)
);

It shows the error, but on addign STORED, it creates a new table as desired. 

Any other solutions please.

**Reply by GaryAustin1** [reply]
My bad... postgres only does stored generated (I'm used to mysql)
You may want to use a view.

**Reply by GaryAustin1** [reply]
Edit: If you are using the Table UI it appears to not handle generated columns.  It is likely updating all values in the row and generating the default error. 
You can update your rows as long as you don't write the generated column in your update.  It is supposed to recalculate on updates.  That might be the cause of your error.

**Reply by compasssites** [reply]
> You can update your rows as long as you don't write the generated column in your update. It is supposed to recalculate on updates. That might be the cause of your error.

It doesnt work. When i update the other 2 columns (one and two), the GENERATED column automatically calculates the value and throws the error. This is for existing rows/records. It behaves well with new records. 

I want the user to be able to update the total stock/used stock and the remaining stock should be generated by a simple mathematical operation/formula.

**Reply by GaryAustin1** [reply]
Are you doing that thru the table UI?
How are you updating the the columns?  I verified it works with the SQL editor only updating the two columns used in the calculation.  
I have opened a bug on the Table UI as it tries to update the calc column with its current value too which generates the error.

I have a table with a single row and id, id1 columns.
ALTER TABLE test4 ADD COLUMN calc int8 GENERATED ALWAYS AS (id + id1) STORED;
UPDATE test4 SET id=20,id1=10;
![image](https://user-images.githubusercontent.com/54564956/156037723-f93fdcfe-d437-4c0c-9606-d9d6bc158339.png)
No error and works

**Reply by compasssites** [reply]
Yes, you are right. 👍🏻

1. It updates well with SQL query. 
2. It updates well for a new record even in table UI. 
3. It does not update from table UI for existing records. 
4. It does not update when tried from the API and app form for existing records. 

Will wait for the resolution. 

Meanwhile, any other workaround? 

Also, can a feature request be considered to add simple calulations in a column with formula type.

**Reply by GaryAustin1** [reply]
The API should work as long as you do not list the calc column in your columns to update.  That is eventually a Postgres thing so will not change.  I’m not sure what an app form is and assume that is part of your code.

I doubt the Table UI will get fixed “soon” as you can’t create these fields with the table UI.    But it is open source now so who knows.

**Reply by compasssites** [reply]
Yes, that works. 

I removed that field from the api form and modifying values in other fields updates this field. So, solves my purpose for the time being. But, i guess i have to wait for bug solutions, as any modifications/playing with table UI throws error, as it updates the whole row i guess. 

Thanks a ton. 🙏
